lib/rspec/matchers/expecteds_for_multiple_diffs.rb

```diff
@@ -0,0 +1,118 @@
+      # @param [Hash] options optional keyword arguments, for example for_matcher
+      # @return [RSpec::Matchers::ExpectedsForMultipleDiffs]
+      def add(expected, options={})
+        for_matcher = options.fetch(:for_matcher, NoMatcher.new)
+        expected_list.concat(expected_from(expected, ForMatcher.from(for_matcher)))
+        self
+      end
+
+      # @api private
+      # Returns message with diff(s) appended for provided differ
+      # factory and actual value if there are any
+      # @param [String] message original failure message
+      # @param [Proc] differ_factory brand new differ factory method (or proc, block)
+      # @param [Any] actual value
+      # @return [String]
+      def message_with_diff(message, differ_factory, actual)
+        diff = diffs(differ_factory, actual)
+        message = "#{message}\n#{diff}" unless diff.empty?
+        message
+      end
+
+    protected
+
+      # @api private
+      # Returns list of expected values.
+      # @return [Array<RSpec::Matchers::ExpectedsForMultipleDiffs::Expected>]
+      def expected_list
+        @_expected_list ||= []
+      end
+
+    private
+
+      def diffs(differ_factory, actual)
+        expected_list.map do |expected|
+          expected.diffs(differ_factory, actual)
+        end.inject([], :+).join("\n")
+      end
+
+      def expected_from(value, for_matcher)
+        return value.expected_list if self.class === value
+        [Expected.new(value, for_matcher)]
+      end
+
+      Expected = Struct.new(:value, :for_matcher)
+      # @api private
+      # Represents expected value with possible knowledge of original
+      # matcher
+      class Expected
+        # @api private
+        # Returns diffs for provided differ factory and actual value
+        # @param [Proc] differ_factory brand new differ factory method (or proc, block)
+        # @param [Any] actual value
+        # @return [String]
+        def diffs(differ_factory, actual)
+          diff = differ_factory.call.diff(actual, value)
+          return [] if diff.empty?
+          ["#{for_matcher.diff_headline}#{diff}"]
+        end
+      end
+
+      ForMatcher = Struct.new(:matcher)
+      # @api private
+      # Represents a matcher diff is being built for
+      class ForMatcher
+        # @api private
+        # Wraps provided matcher in ForMatcher. If it is already a
+        # ForMatcher just returns it.
+        # @param [RSpec::Matchers::BuiltIn::BaseMatcher, RSpec::Matchers::ExpectedsForMultipleDiffs::ForMatcher] matcher or for_matcher value
+        # @return [RSpec::Matchers::ExpectedsForMultipleDiffs::ForMatcher]
+        def self.from(matcher)
+          return matcher if self === matcher
+          new(matcher)
+        end
+
+        # @api private
+        # Returns proper diff headline for wrapper matcher with
+        # matcher description included
+        # @return [String]
+        def diff_headline
+          "Diff for (#{matcher.description}):"
+        end
```

Comment:
`matcher.description` can be arbitrarily long -- for example, `include(some_long_4K_length_string).description` will include all 4K characters of the string.  If it's a long string it's going to obscure the diff.  What do you think about truncating the description so that it merely includes the first N characters of the string (with an appropriate `...` if truncated)?  N could be something like 65 or 70 -- whatever ensures the `Diff` line won't be longer than 80 characters.